chore: Use PodTemplates to support ImageJobs

pmengelbert opened the pull request:
This PR was caused by limits on the size of k8s API requests. By default, `kubectl apply` will add a field, `metadata.annotations.kubectl."kubernetes.io/last-applied-configuration"`, which is a JSON representation of the object. This effectively increases the size of the request.

We ran into size limits when adding a `v1` of the eraser API. Because we want to support a multi-version API, the request to create the CRD for imagejobs was essentially doubled by the addition of the new version.

The component of the CRD that took up the most space was the `ImageJob.Spec` field, a `PodTemplateSpec` type. The generated YAML for this type is extremely long, with a lot of documentation. Being a core kubernetes resource, a `PodTemplateSpec` is very long and complex.

This commit leverages the kubernetes resource `PodTemplate` in order to remove the `ImageJob.Spec` field entirely. Instead of using the `ImageJob` to specify configuration of the `eraser` or `collector` pods, that responsibility is offloaded to a `PodTemplate`.

There is a 1:1 ratio between imagejobs and podtemplates. A given imagejob is expected to have a corresponding podtemplate with the same name, in the same namespace as the controller-manager pod. If no such podtemplate exists, an error is returned and the reconciliation loop spins until the podtemplate is found.

The OwnerReference of the podtemplate is set to its corresponding imagejob. When the imagejob is deleted, the podtemplate will be cleaned up as well.

Finally, I made a change allowing the user to disable the immediate triggering of collector jobs. This was done to avoid situations in the tests where we needed to wait for/delete the initial imagejob in order to test imagelists. Commit 7b49359 should be reverted in a later PR, after #568 has been addressed.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Use PodTemplates to support ImageJobs

This PR was caused by limits on the size of k8s API requests. By
default, `kubectl apply` will add a field, `metadata.annotations.kubectl."kubernetes.io/last-applied-configuration"`,
which is a JSON representation of the object. This effectively increases
the size of the request.

We ran into size limits when adding a `v1` of the eraser API. Because we
want to support a multi-version API, the request to create the CRD for
imagejobs was essentially doubled by the addition of the new version.

The component of the CRD that took up the most space was the
`ImageJob.Spec` field, a `PodTemplateSpec` type. The generated YAML for
this type is extremely long, with a lot of documentation. Being a core
kubernetes resource, a `PodTemplateSpec` is very long and complex.

This commit leverages the kubernetes resource `PodTemplate` in order to
remove the `ImageJob.Spec` field entirely. Instead of using the
`ImageJob` to specify configuration of the `eraser` or `collector` pods,
that responsibility is offloaded to a `PodTemplate`.

There is a 1:1 ratio between imagejobs and podtemplates. A given
imagejob is expected to have a corresponding podtemplate with the same
name, in the same namespace as the controller-manager pod. If no such
podtemplate exists, an error is returned and the reconciliation loop
spins until the podtemplate is found.

The OwnerReference of the podtemplate is set to its corresponding
imagejob. When the imagejob is deleted, the podtemplate will be cleaned
up as well.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Address failure in imagelist_rm_images e2e test

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Test without automated collector

This code should be reverted once the timing logic has been sorted out.
See issue #568 for details.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix error in initial trigger logic

Also, remove unnecessary goroutine.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on controllers/imagecollector/imagecollector_controller.go:
```diff
@@ -163,13 +170,15 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-		log.Info("Queued first ImageCollector reconcile")
-	}()
+
+		log.Info("...Queued first ImageCollector reconcile")
```
why `...` prefix? debugging?

pmengelbert replied:
No reason, will remove

sozercan commented on controllers/imagecollector/imagecollector_controller.go:
```diff
@@ -154,7 +155,13 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 
-	go func() {
+	delay := *repeatPeriod
+	if *repeatImmediate {
```
what is this for? is this different than setting repeat period to 0?

pmengelbert replied:
The collect-scan-erase process can either begin immediately or after the specified delay. Our default has been to start immediately, with no option to change that behavior. This change makes it possible to do either.

The only reason I added this to this PR is because the collector pods were messing up the tests. Some tests (for example `imagelist_rm_images`) test the behavior of the imagelist, and have to jump through hoops to not conflict with the collector which begins immediately after deployment. example: https://github.com/Azure/eraser/blob/76aa58d70fb1cce6f2c61f466dd4b34dceb470b5/test/e2e/tests/imagelist_rm_images/eraser_test.go#L51

Ideally I'd like to move to a more robust scheduling system (example: allow the user to specify a cron string).

pmengelbert replied:
> what is this for? is this different than setting repeat period to 0?

Sorry, misunderstood your question. Yes, it's different. Here's how it works:
- If repeatPeriod = 30 min and repeatImmediate = true, then we will run one job at startup, then the next job 30 minutes after that
- If repeatPeriod = 30 min and repeatImmediate = false, then we will wait 30 mins after startup to run the first job

On the other hand, setting repeatPeriod to 0 would start immediately and repeat endlessly with no delay

sozercan replied:
I see, `repeat` seems a bit misleading. perhaps we can name it something like `schedule-immediate`, `queue-immediate` or `erase-immediate` (or anything else you can think of)?

pmengelbert replied:
done

sozercan commented on controllers/imagecollector/imagecollector_controller.go:
```diff
@@ -233,75 +242,74 @@ func (r *Reconciler) createImageJob(ctx context.Context, req ctrl.Request, argsC
+					},
+					Resources: corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							"cpu":    resource.MustParse("7m"),
```
are we hardcoding these?

pmengelbert replied:
These have been hardcoded for some time, the code just moved due to indentation changes, etc.

sozercan replied:
is this part of the config we talked about before?

pmengelbert replied:
It should be. I'll leave a link to this comment in the configmap issue #446

sozercan commented on controllers/imagecollector/imagecollector_controller.go:
```diff
@@ -163,13 +170,15 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 	return nil
 }
 
 //+kubebuilder:rbac:groups="batch",resources=jobs,verbs=get;list;create;delete;watch
```
can we remove this or is this used?

pmengelbert replied:
We can remove

sozercan replied:
i still see this, do we want to remove in this pr?

pmengelbert replied:
For clarity, you meant the kubebuilder comment? Or the log.Info? Both have been removed.

sozercan replied:
sorry, should have clarified. I meant kubebuilder marker

sozercan commented on controllers/imagecollector/imagecollector_controller.go:
```diff
@@ -359,18 +367,35 @@ func (r *Reconciler) createImageJob(ctx context.Context, req ctrl.Request, argsC
+	}
+	err = r.Create(ctx, &template)
+	if err != nil {
+		log.Info("Could not create collector PodTemplate")
```
log.Error?

pmengelbert replied:
:+1:

pmengelbert committed:
Address PR comments

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Address more PR comments

- rename `repeatImmediate` to `scheduleImmediate`
- remove kubebuilder comment for k8s `jobs` which we no longer use

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Re-generate manifests

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix typo in e2e test

Mistakenly had the old flag, `--repeat-immediate` instead of the new,
`--schedule-immediate`

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan reviewed (approved):
Thank you! LGTM
